### PR TITLE
Add tendency terms

### PR DESCRIPTION
**Pull request by brian-oneill**
This PR adds functors for tendency terms necessary to solve the shallow water equations. Some arrays utilized by these functors have also been added to the HorzMesh class. Functors for all the terms in the momentum and layer thickness equations are included. Ran tests successfully on Chrysalis and Perlmutter CPU & GPU.

Checklist
* [x] Documentation:
  * [x] Design document has been generated and added to the docs
  * [x] User's Guide has been updated
  * [x] Developer's Guide has been updated
  * [x] Documentation has been [built locally](https://e3sm-project.github.io/Omega/develop/devGuide/BuildDocs.html) and changes look as expected
* [x] Testing
  * [x] CTest unit tests for new features have been added per the approved design. 
  * [x] Unit tests have passed. Please provide a relevant CDash build entry for verification.

### CODE REVIEW

**Comment by mark-petersen** [comment]
@brian-oneill please provide the files you used to run your tests. ctest is looking for the files
```
test/OmegaPlanarMesh.nc   test/OmegaSphereMesh.nc
```
and I gave it some standard files but it fails.

**Comment by xylar** [comment]
@mark-petersen, the 3 meshes in https://web.lcrc.anl.gov/public/e3sm/polaris/ocean/omega_ctest/ are what I use in the polaris ctest utility. Try those.

**Comment by xylar** [comment]
Specifically:
```
cd test
wget https://web.lcrc.anl.gov/public/e3sm/polaris/ocean/omega_ctest/ocean.QU.240km.151209.nc
ln -s ocean.QU.240km.151209.nc OmegaMesh.nc
wget https://web.lcrc.anl.gov/public/e3sm/polaris/ocean/omega_ctest/cosine_bell_icos480_initial_state.230220.nc
ln -s cosine_bell_icos480_initial_state.230220.nc OmegaSphereMesh.nc
wget https://web.lcrc.anl.gov/public/e3sm/polaris/ocean/omega_ctest/PlanarPeriodic48x48.nc
ln -s PlanarPeriodic48x48.nc OmegaPlanarMesh.nc
```

File: components/omega/src/ocn/TendencyTerms.h [hunk 1]
```diff
@@ -0,0 +1,424 @@
...
+
+      const I4 KStart         = KChunk * VecLength;
+      Real VortTmp[VecLength] = {0};
+
+      for (int J = 0; J < NEdgesOnEdge(IEdge); ++J) {
+         I4 JEdge = EdgesOnEdge(IEdge, J);
+         for (int KVec = 0; KVec < VecLength; ++KVec) {
+            const I4 K    = KStart + KVec;
+            Real NormVort = (NormRVortEdge(IEdge, K) + NormFEdge(IEdge, K) +
+                             NormRVortEdge(JEdge, K) + NormFEdge(JEdge, K)) * 0.5;
```
Comment (mwarusz) [inline components/omega/src/ocn/TendencyTerms.h:76]:
```suggestion
                             NormRVortEdge(JEdge, K) + NormFEdge(JEdge, K)) * 0.5_Real;
```

File: components/omega/src/ocn/TendencyTerms.h [hunk 2]
```diff
@@ -0,0 +1,424 @@
...
+   KOKKOS_FUNCTION void operator()(const Array2DReal &Tend,
+                                   I4 IEdge,
+                                   I4 KChunk,
+                                   const Array2DR8 &HCell
+                                   ) const {
+      
+      const I4 KStart      = KChunk * VecLength;
+      const I4 ICell0      = CellsOnEdge(IEdge, 0);
+      const I4 ICell1      = CellsOnEdge(IEdge, 1);
+      const Real InvDcEdge = 1. / DcEdge(IEdge);
```
Comment (mwarusz) [inline components/omega/src/ocn/TendencyTerms.h:145]:
```suggestion
      const Real InvDcEdge = 1._Real / DcEdge(IEdge);
```

File: components/omega/src/ocn/TendencyTerms.h [hunk 3]
```diff
@@ -0,0 +1,424 @@
...
+
+      const I4 KStart = KChunk * VecLength;
+      const I4 ICell0 = CellsOnEdge(IEdge, 0);
+      const I4 ICell1 = CellsOnEdge(IEdge, 1);
+
+      const I4 IVertex0 = VerticesOnEdge(IEdge, 0);
+      const I4 IVertex1 = VerticesOnEdge(IEdge, 1);
+
+      const Real DcEdgeInv = 1. / DcEdge(IEdge);
+      const Real DvEdgeInv = 1. / DvEdge(IEdge);
```
Comment (mwarusz) [inline components/omega/src/ocn/TendencyTerms.h:183]:
```suggestion
      const Real DcEdgeInv = 1._Real / DcEdge(IEdge);
      const Real DvEdgeInv = 1._Real / DvEdge(IEdge);
```

File: components/omega/src/ocn/TendencyTerms.h [hunk 4]
```diff
@@ -0,0 +1,424 @@
...
```
Comment (mwarusz) [inline components/omega/src/ocn/TendencyTerms.h:231]:
```suggestion
      const Real DcEdgeInv = 1._Real / DcEdge(IEdge);
      const Real DvEdgeInv = 1._Real / DvEdge(IEdge);
```

File: components/omega/src/ocn/TendencyTerms.h [hunk 5]
```diff
@@ -0,0 +1,424 @@
...
+      for (int J = 0; J < NEdgesOnCell(ICell); ++J) {
+         const I4 JEdge = EdgesOnCell(ICell, J);
+
+         const I4 JCell0 = CellsOnEdge(JEdge, 0);
+         const I4 JCell1 = CellsOnEdge(JEdge, 1);
+
+         for (int KVec = 0; KVec < VecLength; ++KVec) {
+            const I4 K = KStart + KVec;
+            const Real NormTrEdge = 
+                (NormTrCell(L, JCell0, K) + NormTrCell(L, JCell1, K)) * 0.5;
```
Comment (mwarusz) [inline components/omega/src/ocn/TendencyTerms.h:288]:
```suggestion
                (NormTrCell(L, JCell0, K) + NormTrCell(L, JCell1, K)) * 0.5_Real;
```

File: components/omega/src/ocn/TendencyTerms.h [hunk 6]
```diff
@@ -0,0 +1,424 @@
...
+   KOKKOS_FUNCTION void operator()(const Array3DReal &Tend,
+                                   I4 L,
+                                   I4 ICell,
+                                   I4 KChunk,
+                                   const Array3DR8 &NormTrCell,
+                                   const Array2DR8 &HMeanEdge
+                                   ) const {
+
+      const I4 KStart  = KChunk * VecLength;
+      Real InvAreaCell = 1. / AreaCell(ICell);
```
Comment (mwarusz) [inline components/omega/src/ocn/TendencyTerms.h:324]:
```suggestion
      Real InvAreaCell = 1._Real / AreaCell(ICell);
```

File: components/omega/src/ocn/TendencyTerms.h [hunk 7]
```diff
@@ -0,0 +1,424 @@
...
+
+      Real DiffTmp[VecLength] = {0};
+
+      for (int J = 0; J < NEdgesOnCell(ICell); ++J) {
+         const I4 JEdge = EdgesOnCell(ICell, J);
+
+         const I4 JCell0 = CellsOnEdge(JEdge, 0);
+         const I4 JCell1 = CellsOnEdge(JEdge, 1);
+
+         const Real InvDcEdge = 1. / DcEdge(JEdge);
```
Comment (mwarusz) [inline components/omega/src/ocn/TendencyTerms.h:334]:
```suggestion
         const Real InvDcEdge = 1._Real / DcEdge(JEdge);
```

File: components/omega/src/ocn/TendencyTerms.h [hunk 8]
```diff
@@ -0,0 +1,424 @@
...
+                                   I4 L,
+                                   I4 ICell,
+                                   I4 KChunk,
+                                   const Array2DR8 &VEdge,
+                                   const Array3DR8 &NormTrCell,
+                                   const Array2DR8 &HFluxEdge
+                                   ) const {
+
+      const I4 KStart  = KChunk * VecLength;
+      Real InvAreaCell = 1. / AreaCell(ICell);
```
Comment (mwarusz) [inline components/omega/src/ocn/TendencyTerms.h:275]:
```suggestion
      Real InvAreaCell = 1._Real / AreaCell(ICell);
```

File: components/omega/src/ocn/TendencyTerms.h [hunk 9]
```diff
@@ -0,0 +1,424 @@
...
+
+   KOKKOS_FUNCTION void operator()(const Array3DReal &Tend,
+                                   I4 L,
+                                   I4 ICell,
+                                   I4 KChunk,
+                                   const Array3DR8 &TrDel2Cell
+                                   ) const {
+
+      const I4 KStart  = KChunk * VecLength;
+      Real InvAreaCell = 1. / AreaCell(ICell);
```
Comment (mwarusz) [inline components/omega/src/ocn/TendencyTerms.h:381]:
```suggestion
      Real InvAreaCell = 1._Real / AreaCell(ICell);
```

File: components/omega/src/ocn/TendencyTerms.h [hunk 10]
```diff
@@ -0,0 +1,424 @@
...
+
+      Real HypTmp[VecLength] = {0};
+
+      for (int J = 0; J < NEdgesOnCell(ICell); ++J) {
+         const I4 JEdge = EdgesOnCell(ICell, J);
+
+         const I4 JCell0 = CellsOnEdge(JEdge, 0);
+         const I4 JCell1 = CellsOnEdge(JEdge, 1);
+
+         const Real InvDcEdge = 1. / DcEdge(JEdge);
```
Comment (mwarusz) [inline components/omega/src/ocn/TendencyTerms.h:391]:
```suggestion
         const Real InvDcEdge = 1._Real / DcEdge(JEdge);
```

File: components/omega/test/ocn/TendencyTermsTest.cpp [hunk 11]
```diff
@@ -0,0 +1,660 @@
...
+
+   if (!isApprox(PotVortHAdvErrors.L2, Setup.ExpectedPVErrorL2, RTol)) {
+      Err++;
+      LOG_ERROR("TendencyTermsTest: PotVortHAdv L2 FAIL");
+   }
+
+   if (Err == 0) {
+      LOG_INFO("TendencyTermsTest: PotVortHAdv PASS");
+   }
+
```
Comment (mwarusz) [inline components/omega/test/ocn/TendencyTermsTest.cpp:316]:
```suggestion
return Err;
```

**Review by mwarusz — CHANGES_REQUESTED** [review]
I know that this PR is still a work in progress, but I had a couple of suggestion based in my initial reading.

File: components/omega/doc/userGuide/TendencyTerms.md [hunk 12]
```diff
@@ -0,0 +1,38 @@
...
+Tendencies for updating state variables are computed using functors, which are
+class objects that can be called like functions. These functors are used within
+Kokkos parallel loops to compute the full tendency arrays. The following
+tendency terms are currently implemented:
+| Name | Description |
+| ---------------- | ---------------- |
+| ThicknessFluxDivOnCell | divergence of layer thickness flux, defined at cell center
+| PotentialVortHAdvOnEdge | horizontal advection of potential vorticity, defined on edges
+| KEGradOnEdge | gradient of kinetic energy, defined on edges
+| SSHGradONEdge | gradient of sea-surface height, multipled of gravitational acceleration, defined on edges
```
Comment (mark-petersen) [inline components/omega/doc/userGuide/TendencyTerms.md:14]:
```suggestion
| SSHGradOnEdge | gradient of sea-surface height, multiplied by gravitational acceleration, defined on edges
```

**Review by hyungyukang — APPROVED** [review]
@brian-oneill , Everything looks good to me. Thank you for your efforts on this. I've tested this PR by implementing a shallow water model, where the tendency terms are computed by this PR. As expected, the convergence rate with horizontal resolutions exactly follows the second-order slope (tested with the nonlinear manufactured solution), indicating that all computations from this PR, as well as auxiliary variable computations (#90), are accurate:

image

I'm approving this PR, but there are two things I would liked to comment for the future reference:
- The gravity acceleration is now hardcoded as 9.80665, but it should be changed by Config or the upper-level driver in the future.
- In `SSHGradOnEdge`, `HCell` is actually `h-b`. The `BottomDepth (b)` could be included in the `SSHGradOnEdge`, or the total depth can be updated before computing `SSHGradOnEdge`.

File: components/omega/src/ocn/HorzMesh.cpp [hunk 13]
```diff
@@ -559,6 +565,51 @@ void HorzMesh::computeEdgeSign() {
...
+void HorzMesh::setMasks(int NVertLevels) {
+
+   EdgeMask = Array2DR8("EdgeMask", NEdgesSize, NVertLevels);
+
+   OMEGA_SCOPE(o_EdgeMask, EdgeMask);
+
+   parallelFor(
+       {NEdgesAll}, KOKKOS_LAMBDA(int Edge) {
+          for (int K = 0; K < NVertLevels; ++K) {
+             o_EdgeMask(Edge, K) = 1.0;
```
Comment (philipwjones) [inline components/omega/src/ocn/HorzMesh.cpp:581]: These o_ arrays should follow the naming conventions using a CamelCase name starting with a capital letter.

File: components/omega/test/ocn/TendencyTermsTest.cpp [hunk 14]
```diff
@@ -0,0 +1,676 @@
...
+   if (Err == 0) {
+      LOG_INFO("TendencyTermsTest: Successful completion");
+   }
+
+   finalizeTendTest();
+
+} // end tendencyTermsTest
+
+int main(int argc, char *argv[]) {
```
Comment (philipwjones) [inline components/omega/test/ocn/TendencyTermsTest.cpp:665]: With recent changes to the testing, we require main to actually return the error code, so you'll want to have tendencyTermsTest to return Err and then main return that Err as an exit code. This was to catch certain errors that didn't create a FAIL but caused the code to exit.

**Review by philipwjones — APPROVED** [review]
I have not checked correctness, but looks like others have. Just a couple of minor things to change.

I suspect we are going to need to break this up eventually to prevent it from becoming a single 200k LOC routine, but that can maybe be done in a later PR.

**Review by mark-petersen — APPROVED** [review]
Approving based on ctest results and @hyungyukang's testing above.

**Review by mwarusz — APPROVED** [review]
LGTM

**Comment by kieran-ringel** [comment]
image
image
image
I've tested this PR by adding C++ timers around the ParallelFor for each tendency term and a loop to perform the ParallelFor 10 times to get larger timing values. 
`   auto t1 = std::chrono::system_clock::now();
   for(int i = 0; i < 10; i++){
   parallelFor(
      {Mesh->NCellsOwned, NVertLevels}, KOKKOS_LAMBDA(int ICell, int KLevel) {
         ThickFluxDivOnC(NumThickFluxDiv, ICell, KLevel, ThickFluxEdge);
      });
   }
   auto t2 = std::chrono::system_clock::now();
   std::chrono::duration<double> ellapsed = t2 - t1;
   std::cout << "ThickFluxDivTiming " << ellapsed.count() << "\n";`
Convergence tests were done by getting the L2 and LInf error values, these test did not loop over the ParallelFor additional times.

**Comment by kieran-ringel** [comment]
To confirm the resulting values in the exact and numeric results I output the .nc file, I did this using Omega's IO.
`/*
   OMEGA::Decomp *DefDecomp = OMEGA::Decomp::getDefault();

  std::vector<int> OffsetEdge(Mesh->NEdgesSize * NVertLevels, -1);
  for (int Edge = 0; Edge < Mesh->NEdgesOwned; ++Edge) {
        int GlobalEdgeAdd = DefDecomp->EdgeIDH(Edge) - 1;
        for (int k = 0; k < NVertLevels; ++k) {
           int VectorAdd         = Edge * NVertLevels + k;
           OffsetEdge[VectorAdd] = GlobalEdgeAdd * NVertLevels + k;
        }
     }


  int DecompEdgeR8;
  std::vector<int> EdgeDims{DefDecomp->NEdgesGlobal, NVertLevels};
  int EdgeArraySize = Mesh->NEdgesSize * NVertLevels;
  OMEGA::IO::createDecomp(DecompEdgeR8, OMEGA::IO::IOTypeR8, 2,
                                   EdgeDims, EdgeArraySize, OffsetEdge,
                                   OMEGA::IO::DefaultRearr);


  int OutFileID;
  OMEGA::IO::openFile(
        OutFileID, "ThickFluxTest.nc", OMEGA::IO::ModeWrite, OMEGA::IO::FmtDefault,
        OMEGA::IO::IfExists::Replace, OMEGA::IO::Precision::Double);


  int DimEdgeID;
  int DimVertID;
  OMEGA::IO::defineDim(OutFileID, "NVertLevels", NVertLevels, DimVertID);
  OMEGA::IO::defineDim(OutFileID, "NEdges", DefDecomp->NEdgesGlobal, DimEdgeID);


  int VarIDEdgeExact;
  int EdgeDimIDs[2] = {DimEdgeID, DimVertID};
  OMEGA::IO::defineVar(OutFileID, "Exact", OMEGA::IO::IOTypeR8, 2,
                                EdgeDimIDs, VarIDEdgeExact);
    
  int VarIDEdgeComputed;
  OMEGA::IO::defineVar(OutFileID, "Computed", OMEGA::IO::IOTypeR8, 2,
                                EdgeDimIDs, VarIDEdgeComputed);
*/
   // Compute exact result
   Array2DReal ExactThickFluxDiv("ExactThickFluxDiv", Mesh->NCellsOwned,
                                 NVertLevels);

   //OMEGA::R8 Fill = -1.23456789e30;
   //OMEGA::IO::writeArray(ExactThickFluxDiv.data(), Mesh->NEdgesSize * NVertLevels, &Fill, OutFileID, DecompEdgeR8, VarIDEdgeExact);
OMEGA::IO::closeFile(OutFileID);`

**Comment by mark-petersen** [comment]
I tested this branch one more time with CPUs and GPUs on perlmutter. Here are my instructions:

<details>
<summary> CPU on perlmutter </summary> <p>

```
cd /global/homes/m/mpeterse/repos/omega/develop
git fetch
git reset --hard origin/develop
git submodule update --init --recursive externals/ekat externals/scorpio cime
cd components/omega/

module load cmake
DIR=240715_omega_cpu
mkdir ${PSCRATCH}/runs/$DIR
cd !$

rm -rf build
mkdir build
cd build
export PARMETIS_ROOT=/global/cfs/cdirs/e3sm/software/polaris/pm-cpu/spack/dev_polaris_0_3_0_gnu_mpich/var/spack/environments/dev_polaris_0_3_0_gnu_mpich/.spack-env/view
cmake \
   -DOMEGA_CIME_COMPILER=gnu  \
   -DOMEGA_CIME_MACHINE=pm-cpu \
   -DOMEGA_PARMETIS_ROOT=${PARMETIS_ROOT}\
   -DOMEGA_BUILD_TEST=ON \
   -Wno-dev \
   -S /global/homes/m/mpeterse/repos/omega/pr_tendterms_boneill/components/omega -B .
./omega_build.sh

# linking:
cd test
ln -isf /global/homes/m/mpeterse/meshes/omega/O*nc .
# note, original files from here:
# wget https://web.lcrc.anl.gov/public/e3sm/polaris/ocean/omega_ctest/ocean.QU.240km.151209.nc
# ln -s ocean.QU.240km.151209.nc OmegaMesh.nc
# wget https://web.lcrc.anl.gov/public/e3sm/polaris/ocean/omega_ctest/cosine_bell_icos480_initial_state.230220.nc
# ln -s cosine_bell_icos480_initial_state.230220.nc OmegaSphereMesh.nc
# wget https://web.lcrc.anl.gov/public/e3sm/polaris/ocean/omega_ctest/PlanarPeriodic48x48.nc
# ln -s PlanarPeriodic48x48.nc OmegaPlanarMesh.nc

# run test:
salloc --nodes 1 --qos interactive --time 01:00:00 --constraint cpu --account=m4572 # or e3sm
DIR=240715_omega_cpu
cd ${PSCRATCH}/runs/${DIR}/build

./omega_ctest.sh

also:
./omega_ctest.sh -R TEND_PLANE_TEST

# output in Testing/Temporary/LastTest.log
```
</details>

<details>
<summary> GPU on perlmutter </summary> <p>

```

DIR=240715_omega_gpu
mkdir ${PSCRATCH}/runs/$DIR
cd !$

rm -rf build
mkdir build
cd build
PARMETIS_ROOT=/global/cfs/cdirs/e3sm/software/polaris/pm-gpu/spack/dev_polaris_0_3_0_gnugpu_mpich/var/spack/environments/dev_polaris_0_3_0_gnugpu_mpich/.spack-env/view
module load cmake
cmake \
   -DOMEGA_CIME_COMPILER=gnugpu  \
   -DOMEGA_CIME_MACHINE=pm-gpu \
   -DOMEGA_PARMETIS_ROOT=${PARMETIS_ROOT}\
   -DOMEGA_BUILD_TEST=ON \
   -Wno-dev \
   -S /global/homes/m/mpeterse/repos/omega/pr_tendterms_boneill/components/omega -B .
./omega_build.sh

# linking:
cd test
ln -isf /global/homes/m/mpeterse/meshes/omega/O*nc .
# note, original files from here:
# wget https://web.lcrc.anl.gov/public/e3sm/polaris/ocean/omega_ctest/ocean.QU.240km.151209.nc
# ln -s ocean.QU.240km.151209.nc OmegaMesh.nc
# wget https://web.lcrc.anl.gov/public/e3sm/polaris/ocean/omega_ctest/cosine_bell_icos480_initial_state.230220.nc
# ln -s cosine_bell_icos480_initial_state.230220.nc OmegaSphereMesh.nc
# wget https://web.lcrc.anl.gov/public/e3sm/polaris/ocean/omega_ctest/PlanarPeriodic48x48.nc
# ln -s PlanarPeriodic48x48.nc OmegaPlanarMesh.nc

# run test:
salloc --nodes 4 --qos interactive --time 01:00:00 --constraint gpu --tasks-per-node=2 --gpus-per-task 1 --account=m4572_g # or e3sm_g
DIR=240715_omega_gpu
cd ${PSCRATCH}/runs/${DIR}/build

./omega_ctest.sh

```

Passes all 17 tests on both.
</details>

**Pull request merged**